pandas/io/formats/excel.py

```diff
@@ -472,6 +472,7 @@ def __init__(
         self.na_rep = na_rep
         if not isinstance(df, DataFrame):
             self.styler = df
+            self.styler._compute()
```

Comment:
this clause is only called when you do `styler.to_excel`. It needs `_compute` to calculate the styles added. Previously, this was done in `_generate_body` below, but, since, it is now needed for both body and headers I brought it to the obvious point: initialisation, to avoid having to unnecessarily do it twice.

Reply:
could maybe move this to `styler.to_excel` and maybe do the pre-calc there before passing  to ExcelFormatter. Might be more sensible like that.

Reply:
> could maybe move this to styler.to_excel and maybe do the pre-calc there before passing to ExcelFormatter. Might be more sensible like that.

I tried it and it didn't work. The new tests failed when using a custom ExcelFormatter, the styles computation needs to be done within ExcelFormatter.